design-proposals/descheduler-support.md

```diff
@@ -0,0 +1,97 @@
+
+This document proposes addition of functionality to KubeVirt intended to support the descheduler's operation on `virt-launcher` pods. 
+
+## Motivation
+Cluster administrators want to have a descheduler running on their cluster.
+The descheduler should be able to operate on `virt-launcher` pods.
+
+## Goals
+1. Enable KubeVirt to work on existing K8s clusters that already have a descheduler deployed.
+2. Enable cluster admins to introduce a descheduler to an existing cluster with KubeVirt deployed.
+
+## Non Goals
+- The Kubernetes eviction API will not be changed.
+- The descheduler is expected to respect the agreement described bellow.
+
+## Definition of Users
+- Cluster administrator
+
+## User Stories
+- As a cluster admin running a K8s cluster with a descheduler, I want to deploy and use KubeVirt.
+- As a cluster admin running a K8s cluster with KubeVirt, I want to deploy a descheduler that will correctly operate on my VMs.
+- As a cluster admin I want to enable cluster-wide KubeVirt's descheduler support.
+
+## Repos
+1. kubevirt/kubevirt
+2. kubevirt/hyperconverged-cluster-operator
+
+# Design
+## Descheduler Requirements
+1. In order for the descheduler to be able to differentiate `virt-launcher` pods from other pods, all `virt-launcher` pods should be annotated with `descheduler.alpha.kubernetes.io/request-evict-only`.
+2. In cases where the VM will be migrated, eviction requests on `virt-launcher` pods should return 429 HTTP code with a special reason message.
+3. In order for the deschduler to understand that a `virt-launcher` pod is pending migration, it should be annotated with `descheduler.alpha.kubernetes.io/eviction-in-progress` so it will not try to evict it again.
+4. In case the migration failed, the `descheduler.alpha.kubernetes.io/eviction-in-progress` annotation will be removed from the source `virt-launcer` so the descheduler will understand that it could try evicting it again.
+
+## Feature Gate
+A `DeschedulerSupport` feature gate will be introduced in order to enable this feature.
+
+## Changes to Pod Eviction Webhook
+Currently, the descheduler observes the one of the following responses:
+
+| Eviction Strategy     | Is VMI migratable | Does Webhook approve eviction | Does PDB allow eviction | Response                         |
+|-----------------------|-------------------|-------------------------------|-------------------------|----------------------------------|
+| None                  | N/A               | True                          | True                    | 200 - Eviction granted           |
+| LiveMigrate           | True              | True                          | False                   | 429 - Eviction blocked by PDB    |
+| LiveMigrate           | False             | False                         | False                   | 429 - Eviction denied by webhook |
```

Comment:
@orelmisan @fossedihelm @ingvagabund 

It seems strange to me that we return the same response for VMIs that won't be evicted as we do for those that will be evicted. According to this [response](https://github.com/kubevirt/community/pull/258/files#r1489367421), a 429 response means that the launcher pod will be counted as an evicted pod. This is odd because, in cases where the VMI is not migratable, the launcher pod won't be evicted and therefore shouldn't be counted as such.

Am I missing something?

Reply:
The returned structure has an error code and message fields.
For cases where the VMI is evacuated, the message field is populated with a dedicated message.
Please see:
- https://github.com/kubevirt/kubevirt/pull/11898

- https://github.com/kubevirt/kubevirt/blob/main/pkg/virt-api/webhooks/validating-webhook/admitters/pod-eviction-admitter_test.go

Reply:
Following https://github.com/kubevirt/kubevirt/pull/11532 The descheduler can understand whether a virt-launcher pod will be evacuated or not - based on the error message.

Reply:
I hope it does. It sounds odd to me to rely on an error message.

Reply:
Prior to that PR, the descheduler couldn't tell if a virt-launcher pod was evacuated in the background.
It only saw error code 429, due to the PDBs protecting the pods and assumed the pods will not move.